File: src/lib/Core/DB/Role/Role.cpp
#pragma once
#include "Role.h"
Role* DB::RoleConstruct::Create(RoleName name) {
    switch (name)
    {
    case RoleName::阿贝多:
        return new 阿贝多();
    case RoleName::埃洛伊:
        return new 埃洛伊();
    case RoleName::安柏:
        return new 安柏();
    case RoleName::荒泷一斗:
        return new 荒泷一斗();
    case RoleName::芭芭拉:
        return new 芭芭拉();
    case RoleName::北斗:
        return new 北斗();
    case RoleName::班尼特:
        return new 班尼特();
    case RoleName::重云:
        return new 重云();
    case RoleName::迪卢克:
        return new 迪卢克();
    case RoleName::迪奥娜:
        return new 迪奥娜();
    case RoleName::优菈:
        return new 优菈();
    case RoleName::菲谢尔:
        return new 菲谢尔();
    case RoleName::甘雨:
        return new 甘雨();
    case RoleName::五郎:
        return new 五郎();
    case RoleName::胡桃:
        return new 胡桃();
    case RoleName::琴:
        return new 琴();
    case RoleName::枫原万叶:
        return new 枫原万叶();
    case RoleName::凯亚:
        return new 凯亚();
    case RoleName::神里绫华:
        return new 神里绫华();
    case RoleName::神里绫人:
        return new 神里绫人();
    case RoleName::刻晴:
        return new 刻晴();
    case RoleName::可莉:
        return new 可莉();
    case RoleName::九条裟罗:
        return new 九条裟罗();
    case RoleName::丽莎:
        return new 丽莎();
    case RoleName::莫娜:
        return new 莫娜();
    case RoleName::凝光:
        return new 凝光();
    case RoleName::诺艾尔:
        return new 诺艾尔();
    case RoleName::七七:
        return new 七七();
    case RoleName::雷电将军:
        return new 雷电将军();
    case RoleName::雷泽:
        return new 雷泽();
    case RoleName::罗莎莉亚:
        return new 罗莎莉亚();
    case RoleName::珊瑚宫心海:
        return new 珊瑚宫心海();
    case RoleName::早柚:
        return new 早柚();
    case RoleName::申鹤:
        return new 申鹤();
    case RoleName::砂糖:
        return new 砂糖();
    case RoleName::达达利亚:
        return new 达达利亚();
    case RoleName::托马:
        return new 托马();
    case RoleName::温迪:
        return new 温迪();
    case RoleName::香菱:
        return new 香菱();
    case RoleName::魈:
        return new 魈();
    case RoleName::行秋:
        return new 行秋();
    case RoleName::辛焱:
        return new 辛焱();
    case RoleName::八重神子:
        return new 八重神子();
    case RoleName::烟绯:
        return new 烟绯();
    case RoleName::宵宫:
        return new 宵宫();
    case RoleName::云堇:
        return new 云堇();
    case RoleName::钟离:
        return new 钟离();
    default:
        throw "";
    }
}

File: CodeCoder/src/Role.ts
import * as genshindb from "genshin-db"
import * as fs from "fs"
const RolePath = "..\\src\\lib\\Core\\DB\\Role\\"
const ElementMap: any = {}
const SkillDefName = ['A','E','Q','SP']
let RoleNames:any[] = [];
genshindb.characters('names', { matchCategories: true }).map(name => {
    let character = genshindb.characters(name)
    let characterC = genshindb.characters(name, { resultLanguage: genshindb.Languages.ChineseSimplified })
    let data = genshindb.talents(name, { resultLanguage: genshindb.Languages.ChineseSimplified })
    let star = genshindb.constellations(name, { resultLanguage: genshindb.Languages.ChineseSimplified })
    // console.log(character);

    if (data && character && star) {
        let { gender: 性别, element: 属性, weapontype: 武器, substat: 突破属性 } = character;
        let 技能表 = [data.combat1, data.combat2, data.combat3];
        
        let 被动表 = [data.passive1, data.passive2]
        data.passive3 && 被动表.push(data.passive3)
        data.passive4 && 被动表.push(data.passive4)
        data.combatsp && 被动表.push(data.combatsp)
        let stat = character.stats(90);
        let 命座 = [star.c1, star.c2, star.c3, star.c4, star.c5, star.c6]
        let 名字 = data.name;
        ElementMap[属性] = ""
        let H =
            `\uFEFF#pragma once
        #include "../Role.h"
        namespace DB {
            namespace RoleConstruct {        
                class ${名字} : public Advanced::Role
                {
                public:
                    ~${名字}();
                    ${名字}(u32*);
                    void A(Role*, u32);
                    void E(Role*, u32);
                    void Q(Role*, u32);
                };
            }
        }
        `;
        let CPP =
            `\uFEFF#pragma once
        #include "${名字}.h"
        using namespace DB::RoleConstruct;
        using namespace Atom;
        using namespace Atom::Enum;
        using namespace Advanced;
        using namespace Professional;
        //TODO:AUTO${名字}.cpp
        const double HP =  ${stat.hp || 0};
        const double ATK =  ${stat.attack || 0};
        const double DEF =  ${stat.defense || 0};
        //${characterC?.substat}
        const double OTHER = ${stat.specialized || 0};
        
        ${名字}::~${名字}() {
        }
        ${名字}::${名字}(u32* frameCur):Role(frameCur) {
            BaseObject* baseData = new BaseObject();
            static Attr
			* HelathAttr = new Attr(HP, DOUBLEZERO, DOUBLEZERO),
			* AtkAttr = new Attr(ATK, DOUBLEZERO, DOUBLEZERO),
			* DefAttr = new Attr(DEF, DOUBLEZERO, DOUBLEZERO),
			* CritRateAttr = new Attr(0.05, DOUBLEZERO, DOUBLEZERO),
			* CritDamageAttr = new Attr(0.5, DOUBLEZERO, DOUBLEZERO);
            baseData->SetAttr(HelathAttr,AttrType::Helath);
		    baseData->SetAttr(AtkAttr,AttrType::Atk);
		    baseData->SetAttr(DefAttr,AttrType::Def);
		    baseData->SetAttr(CritRateAttr,AttrType::CritRate);
		    baseData->SetAttr(CritDamageAttr,AttrType::CritDamage);
            this->rawCharacter = new Character(baseData, CharacterGender::${性别}, CharacterGroup::Tivat, ElementType::${属性}, WeaponType::${武器});
            delete baseData;
        };        

        /*
        ${被动表.map((被动, 序号) => {
                return `
        ****${序号 + 1}:${被动.name}
        ****************
        ${被动.info.split("\n").map(i => {
                    return `
            ${i}`
                }).join("").replace(/\*\*/g, "❤")} 

        ****************
        `
            }).join("\n")}
        */
        ${技能表.map((技能, 序号) => {
                return `
            /*******${技能.name}
             * ${技能.info.split("\n").map(info=>`
             * ${info}`).join("")}
             * ${技能.attributes.labels.map(label => {
                    return `
             * ${label}`
                }).join("")}
            */
            void ${名字}::${SkillDefName[序号]}(Role* role, u32 cmd) {
                const static double** SkillPrama = new const double* [15]{${
                        [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14].map(index=>{
                        return `
                        new const double[${Object.keys(技能.attributes.parameters).length}]{ ${Object.keys(技能.attributes.parameters).map((参数键, 参数序号) => {
                            return 技能.attributes.parameters[参数键][index]
                        }).join(",")
                    }}`
                    })}                    
                };  
                const double* curData = SkillPrama[cmd & 0xFF];
                ////////下面是技能实现   
            };`}).join("\n\n")}
        /*****${star.name}
        ${命座.map((命, 序号) => {
                return `
            ***${序号 + 1}命***
            ${命.name}            
            ${命.effect.split("\n").map(i => {
                    return `${i}`
                }).join("").replace(/\*\*/g, "❤")}
            `
            }).join("")}
        */
         
        `
        let CPPPath = RolePath + `${属性}/${名字}.cpp`
        let HPath = RolePath + `${属性}/${名字}.h`
        // fs.writeFileSync(CPPPath, CPP)
        // fs.writeFileSync(HPath, H)
        fs.stat(CPPPath, (e, stat) => {
            if (stat && stat.isFile()) {
                console.log(!!fs.readFileSync(CPPPath).toString().match(/TODO:AUTO/),名字);
                
                if (!!fs.readFileSync(CPPPath).toString().match(/TODO:AUTO/)) {
                    fs.writeFileSync(CPPPath, CPP)
                    fs.writeFile(HPath, H, () => { })
                }

            }else{
                fs.writeFileSync(CPPPath, CPP)
                fs.writeFile(HPath, H, () => { })
            }
        })
        // console.log(CPPPath,HPath);       
        RoleNames.push({名字,属性})

    }

});
const RoleH = `\uFEFF#pragma once
#include "../DB.h"

namespace DB {
	namespace RoleConstruct {
        enum class RoleName
		{${RoleNames.map(i=>`
            ${i.名字},`).join("")}
		};
		Advanced::Role* Create(RoleName, u32*);
	}    
}
${RoleNames.map(i=>{
    return `#include "${i.属性}/${i.名字}.h"\n`
}).join("")}
`
const RoleCPP = 
`\uFEFF#pragma once
#include "Role.h"
using namespace Advanced;
Role* DB::RoleConstruct::Create(RoleName name, u32* frameCur) {
    switch (name)
    {${RoleNames.map(i=>`
    case RoleName::${i.名字}:
        return new ${i.名字}(frameCur);`).join("")}
    default:
        throw "";
    }
}
`
// console.log(RoleH, RolePath + "Role.h");


fs.writeFile(RolePath + "Role.h", RoleH, () => { })
fs.writeFile(RolePath + "Role.cpp", RoleCPP, () => { })
